### PR TITLE
[DeviceMesh] Make dtype of mesh tensor from `init_device_mesh()` consistent with directly calling `DeviceMesh()`

### PR DESCRIPTION
**Pull request by wz337**
Currently, mesh tensor from `init_device_mesh()` has a dtype of `torch.int64` while mesh tensor from `DeviceMesh()` would have dtype of `torch.int32`. Making them consistent in this PR. 

DeviceMesh ctor dtype pointer:
https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L217

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @tianyu-l @wconstab @yf225 @chauhang

### CODE REVIEW

**Review by xunnanxu — APPROVED** [review]
LGTM!

File: torch/distributed/device_mesh.py [hunk 1]
```diff
@@ -553,7 +553,7 @@ def init_device_mesh(
                     f"Found len(mesh_dim_names): {len(mesh_dim_names)} and len(mesh_shape):{len(mesh_shape)}.",
                 )
 
-        mesh = torch.arange(math.prod(mesh_shape)).view(mesh_shape)
+        mesh = torch.arange(math.prod(mesh_shape), dtype=torch.int).view(mesh_shape)
```
Comment (wanchaol) [inline torch/distributed/device_mesh.py:556]: I think one additional place need to be change: https://github.com/pytorch/pytorch/blob/main/torch/distributed/device_mesh.py#L215

We need to change this `mesh.detach().cpu()` to sth like `mesh.detach().to(device="cpu", dtype=torch.int)

**Review by wanchaol — APPROVED** [review]
looks good to me! one comment
